src/NodeFinder/FormFieldsFinder.php

```diff
@@ -0,0 +1,172 @@
+                    continue;
+                }
+
+                $addFieldMethodCall = $this->findAddFieldMethodCall($methodCall);
+                if (! $addFieldMethodCall) {
+                    continue;
+                }
+
+                if (! $this->isFormAddFieldMethodCall($addFieldMethodCall, $form)) {
+                    continue;
+                }
+
+                $formFields = $this->addFormField($formFields, $addFieldMethodCall, $methodCall);
+            }
+        }
+        return $formFields;
+    }
+
+    private function findMethodCall(Expression $expression): ?MethodCall
+    {
+        $methodCall = null;
+        if ($expression->expr instanceof MethodCall) {
+            $methodCall = $expression->expr;
+        } elseif ($expression->expr instanceof Assign && $expression->expr->expr instanceof MethodCall) {
+            $methodCall = $expression->expr->expr;
+        }
+        return $methodCall;
+    }
+
+    private function findAddFieldMethodCall(MethodCall $methodCall): ?MethodCall
+    {
+        if ($methodCall->var instanceof Variable) {
+            // skip submit buttons
+            if ($this->nodeTypeResolver->isObjectType(
+                $methodCall,
+                new ObjectType('Nette\Forms\Controls\SubmitButton')
+            )) {
+                return null;
+            }
+            if ($this->nodeTypeResolver->isObjectType($methodCall, new ObjectType('Nette\Forms\Container'))) {
+                return $methodCall;
+            }
+            // skip groups, renderers, translator etc.
+            if ($this->nodeTypeResolver->isObjectType(
+                $methodCall,
+                new ObjectType('Nette\Forms\Controls\BaseControl')
+            )) {
+                return $methodCall;
+            }
+            return null;
+        }
+
+        if ($methodCall->var instanceof MethodCall) {
+            return $this->findAddFieldMethodCall($methodCall->var);
+        }
+        return null;
+    }
+
+    private function isFormAddFieldMethodCall(MethodCall $addFieldMethodCall, Variable $form): bool
+    {
+        $methodCallVariable = $this->findMethodCallVariable($addFieldMethodCall);
+        if ($methodCallVariable === null) {
+            return false;
+        }
+        if ($methodCallVariable->name !== $form->name) {
+            return false;
+        }
+        return true;
+    }
+
+    private function findMethodCallVariable(MethodCall $methodCall): ?Variable
+    {
+        if ($methodCall->var instanceof Variable) {
+            return $methodCall->var;
+        }
+
+        if ($methodCall->var instanceof MethodCall) {
+            return $this->findMethodCallVariable($methodCall->var);
+        }
+        return null;
+    }
+
+    /**
+     * @param FormField[] $formFields
+     * @return FormField[]
+     */
+    private function addFormField(array $formFields, MethodCall $addFieldMethodCall, MethodCall $methodCall): array
```

Comment:
This methods looks like it create a new `FormField`, but the name says it add something (to something else).

Maybe decopule these 2 operations?

Reply:
Yeah but that will introduce another `if ($formField === null) { continue; }`... I've been there. I am trying to lower complexity.

My previous implementation includes reference `&$formFields` but it is not available here.

Reply:
Ok, you're right, these might be hard to grasp.

You've done great job to make it till 10, thanks for that :+1: